File: ACTIVIDADES/Actividad N°3/Cola.py
from collections import deque

class Cola:
    def __init__(self):
        self.elementos = deque()
    
    def enqueue(self, valor):
        """Insertar elemento (enqueue)"""
        self.elementos.append(valor)
        print(f"Insertado: {valor}")
    
    def dequeue(self):
        """Eliminar elemento (dequeue)"""
        if not self.elementos:
            print("Error: Cola vacía")
            return None
        valor = self.elementos.popleft()
        print(f"Eliminado: {valor}")
        return valor
    
    def frente(self):
        """Mostrar el frente"""
        if not self.elementos:
            print("Cola vacía")
            return None
        return self.elementos[0]
    
    def mostrar(self):
        """Recorrer cola"""
        if not self.elementos:
            print("Cola vacía")
            return
        print("Estado de la cola (frente -> final):", end=" ")
        for elemento in self.elementos:
            print(elemento, end=" ")
        print()
    
    def esta_vacia(self):
        return len(self.elementos) == 0
    
    def tamaño(self):
        return len(self.elementos)

# Función principal
def main():
    cola = Cola()
    
    print("=== IMPLEMENTACIÓN DE COLA EN PYTHON ===")
    
    # Insertar elementos: 3, 6, 9, 12
    print("\n1. Insertando elementos:")
    cola.enqueue(3)
    cola.enqueue(6)
    cola.enqueue(9)
    cola.enqueue(12)
    
    print("\nEstado después de insertar:")
    cola.mostrar()
    print(f"Frente actual: {cola.frente()}")
    
    # Eliminar un elemento
    print("\n2. Eliminando un elemento:")
    cola.dequeue()
    
    # Mostrar cola final
    print("\n3. Estado final:")
    cola.mostrar()
    print(f"Frente final: {cola.frente()}")

if __name__ == "__main__":
    main()

File: ACTIVIDADES/Actividad N°3/Cola.cpp
#include <iostream>
#include <queue>
using namespace std;

class Cola {
private:
    queue<int> elementos;
    
public:
    // Insertar elemento (enqueue)
    void enqueue(int valor) {
        elementos.push(valor);
        cout << "Insertado: " << valor << endl;
    }
    
    // Eliminar elemento (dequeue)
    int dequeue() {
        if (elementos.empty()) {
            cout << "Error: Cola vacía" << endl;
            return -1;
        }
        int valor = elementos.front();
        elementos.pop();
        cout << "Eliminado: " << valor << endl;
        return valor;
    }
    
    // Mostrar el frente
    int frente() {
        if (elementos.empty()) {
            cout << "Cola vacía" << endl;
            return -1;
        }
        return elementos.front();
    }
    
    // Recorrer cola (implementación auxiliar)
    void mostrar() {
        if (elementos.empty()) {
            cout << "Cola vacía" << endl;
            return;
        }
        
        queue<int> temp = elementos;
        cout << "Estado de la cola (frente -> final): ";
        while (!temp.empty()) {
            cout << temp.front() << " ";
            temp.pop();
        }
        cout << endl;
    }
    
    bool estaVacia() {
        return elementos.empty();
    }
    
    int tamaño() {
        return elementos.size();
    }
};

// Función principal
int main() {
    Cola cola;
    
    cout << "=== IMPLEMENTACION DE COLA EN C++ ===" << endl;
    
    // Insertar elementos: 3, 6, 9, 12
    cout << "\n1. Insertando elementos:" << endl;
    cola.enqueue(3);
    cola.enqueue(6);
    cola.enqueue(9);
    cola.enqueue(12);
    
    cout << "\nEstado despues de insertar:" << endl;
    cola.mostrar();
    cout << "Frente actual: " << cola.frente() << endl;
    
    // Eliminar un elemento
    cout << "\n2. Eliminando un elemento:" << endl;
    cola.dequeue();
    
    // Mostrar cola final
    cout << "\n3. Estado final:" << endl;
    cola.mostrar();
    cout << "Frente final: " << cola.frente() << endl;
    
    return 0;
}
